api/src/main/java/bio/terra/tanagra/exception/SystemException.java

```diff
@@ -0,0 +1,33 @@
+package bio.terra.tanagra.exception;
+
+/**
+ * Custom exception class for system or internal exceptions. These represent errors that the user
```

Comment:
My thinking here is that there are errors that are understandable to the user (e.g. artifact not found, cohort name uses invalid characters, you don't have access to this underlay) and errors that are not (e.g. BQ is not accepting any requests, our SQL logic got a null ptr exception).

Since I've been overhauling the indexing & query code, I've only hit cases where the details are useless to the user (i.e. Tanagra has a bug, go talk to the admin). I'm hoping that once I get to the artifact storage APIs and implementing access control, a second exception class here will be useful. If it's not, then I'll remove this.

Reply:
I'm not opposed to the idea, but it doesn't appear it's actually serving its purpose here.
1) It's only used in a single place in the PR, but there are are several other types that are used frequently: IOException, UnsupportedOperation, IllegalArgument, ClassCast. That one instance isn't special enough to warrant its own exception. It can be added later if there's actually a need.
2) Remember that users are researchers not programmers. The set of errors they can meaningfully do something about are very limited and will represent a small slice of possible errors. Of the three you listed, the first two are UI bugs outside of rare edge cases so they don't warrant special handling.
3) Errors have to meaningfully transit the API. The third example is useful but the UI needs more information in order to meaningfully handle them. If anything, I would suggest defining a user facing exception instead for specifically cases like that. Consider how Docs displays a special access denied page with a way to request permission. A generic string isn't enough. The response from the API needs a particular response code and machine parsable information that specifies who the owner is or how to request access.